### ISSUE TITLE
Use gson for JSON functionality

### ISSUE THREAD

**Issue by carlschroedl** [happyjack27/autoredistrict#15]
Lots of folks at my work and in the wider software word use Google's gson for (de)serializing to/from JSON. Gson has very sensible, concise default serialization behavior, while still permitting you to customize serialization where you need. Incorporating gson would simplify the existing code base by eliminating most of the serialization logic.

The advantages of gson over current checked-in dependencies are similar to those outlined in  #14.

  1. it has automated tests, and by virtue of being widely-used, is battle-tested against many corner cases
  2. many developers are already familiar with it, so it would lower the barrier to contribution
  3. the project publishes their artifacts in maven central so if we move to a formal dependency management solution like maven, gradle, or ivy, we could stop checking in the jar

What do you think?

**Comment by happyjack27** [comment]
All in all I'm all about lowering the barrier to enter and you already know I love json.

I'm reminded of things like kolmogrov complexity - how do you measure complexity?  I recall a quote by butler Lampson of bell labs: "everything must be made extremely simple, but to do this, one must first understand complexity."

I'm all about making things more accessible.  But I fear inadvertently making it less accessible while trying to make it more.

**Comment by happyjack27** [comment]
All in all I want to try to avoid creating complexity in the name if simplicity.  What is the phrase?   Never use a small word where a diminutive one will suffice.

Above all I want to avoid barriers to entry; barriers to understanding the project.

**Comment by happyjack27** [comment]
But sure in what comment to out this. 

You can always expand upon the core.  I would not be upset by a second repository that was more web-based.  For what I can control i'm more concerned with education.  For instance someone might say "but why a global solution?" And if respond "a globally worse solution means precisely one that makes the average person more unhappy, and is therefore necessarily bad.  A globally good solution is precisely one that is not necessarily bad. Q.E.D.

**Comment by happyjack27** [comment]
Having said all that, I'm quite willing to give you an ear. Just know that any "ad populum" argument will get you quickly ignored.

**Comment by carlschroedl** [comment]
As a new contributor, my biggest barrier to understanding is separating this project's wheat from the chaff. Recall the Unix philosophy - do one thing and do it well. For this tool to be successful we need to focus on assessing and generating representative districts. Why would we check code into this repository that is unrelated to solving that problem? Re-inventing JSON (de)serialization is far from the main goal. Truly novel contributions to that problem space belong in a separate project dedicated to that problem space.

**Comment by carlschroedl** [comment]
Could you please clarify what you mean by "But sure in what comment to out this" ? Might be a phone keyboard error.

**Comment by happyjack27** [comment]
Please disregard my earlier drunken ramblings. :P

**Comment by happyjack27** [comment]
The serialization routine i wrote (which is still a little sloppy), has advantages over gson:
* in-place serialization/deserialization.
* requires less code to serialize/desererialize
* faster / less performance overhead

**Comment by happyjack27** [comment]
i'm not really using it anymore, actually.  except for enabling reading geojson.
other use case i had in mind was .project files but i've stopped maintaining that.

if you have a simple object you actually don't have to do anything (write any code) at all - just extend reflectionjsonobject and it'll use reflection to serialize/deserialize all the members.

if you got something more complex, it'll serialze/deserialize to the collection - you write to that collection before serializing, and read from it after serializing, and everytime it's got a complex object it'll call a function with the name of the object and you instantiate the object and give it to it.  

very simple.  saves a lot of unneccessary coding and objects, and goes a lot faster, because it's in-line.

**Comment by happyjack27** [comment]
looked at gson.  hopelessly bloated. violates the no bloat rule.

**Comment by happyjack27** [comment]
maybe some simple documentation on how to use JsonObject or ReflectionJson Object wouldn't hurt.  at the top of the respective files.

**Comment by carlschroedl** [comment]
Could you please distinguish your "no bloat rule" from the Not Invented Here Anti-Pattern?

**Comment by happyjack27** [comment]
if it's bloated, don't use it.
if there was a better json serializer that did what i wanted and wasn't
bloated, i'd use it.
nothing to do with not invented here, just couldn't find a good alternative.
all you need to make a serializer is extend one class.  i'm not going to
download 100 classes to do something i can do in 1.

**Comment by happyjack27** [comment]
Make new map coloring modes that divide by area - how to normalize it nationally?

Reason is to give a better visual sense

Sent from my iPhone

**Comment by happyjack27** [comment]
Need to cite source of data on notes, so can be reproduced 

Sent from my iPhone

**Comment by happyjack27** [comment]
I needed to convert json to a nested hashmap for an unrelated project.  after searching the web, i found the best fit to my needs was this code that i had written before.  With it, I can do what i want in two lines:

			JSONObject jsonObject = new ReflectionJSONObject<Object>();
			jsonObject.fromJSON(response.toString());

Had I made a constructor, it'd be one line.   (don't know why that didn't occur to me.)

And all the code I had to import was 2 classes (one of which was abstract) and an interface.
